File: benchmarks/tsb/bench_dataframe_value_counts.ts
import { dataFrameValueCounts } from "tsb";
import { DataFrame } from "tsb";
const N = 100_000;
const cats = ["apple", "banana", "cherry", "date", "elderberry"];
const df = new DataFrame({
  fruit: Array.from({ length: N }, (_, i) => cats[i % cats.length]),
  color: Array.from({ length: N }, (_, i) => (i % 3 === 0 ? "red" : i % 3 === 1 ? "yellow" : "purple")),
});
const WARMUP = 3;
const ITERS = 20;
for (let i = 0; i < WARMUP; i++) dataFrameValueCounts(df, { subset: ["fruit", "color"] });
const t0 = performance.now();
for (let i = 0; i < ITERS; i++) dataFrameValueCounts(df, { subset: ["fruit", "color"] });
const total = performance.now() - t0;
console.log(JSON.stringify({ function: "dataframe_value_counts", mean_ms: total / ITERS, iterations: ITERS, total_ms: total }));

File: benchmarks/pandas/bench_dataframe_value_counts.py
import pandas as pd, time, json
N = 100_000
cats = ["apple", "banana", "cherry", "date", "elderberry"]
df = pd.DataFrame({
    "fruit": [cats[i % len(cats)] for i in range(N)],
    "color": ["red" if i % 3 == 0 else "yellow" if i % 3 == 1 else "purple" for i in range(N)],
})
WARMUP = 3
ITERS = 20
for _ in range(WARMUP):
    df.value_counts(subset=["fruit", "color"])
t0 = time.perf_counter()
for _ in range(ITERS):
    df.value_counts(subset=["fruit", "color"])
total = (time.perf_counter() - t0) * 1000
print(json.dumps({"function": "dataframe_value_counts", "mean_ms": total / ITERS, "iterations": ITERS, "total_ms": total}))
